test/support/utils.rb

```diff
@@ -47,41 +47,44 @@ def create_pseudo_terminal(boot_options: "-r debug/run")
 
       ENV['RUBYOPT'] = "-I #{lib}"
       ENV['RUBY_DEBUG_USE_COLORIZE'] = "false"
+      ENV['RUBY_DEBUG_TEST_MODE'] = 'true'
 
       PTY.spawn("#{RUBY} #{boot_options} #{temp_file_path}") do |read, write, pid|
-        quit = false
-        result = nil
-
-        until quit
-          read.expect(/(.*)\n|\(rdbg\)/) do |sentence|
-            debug_print sentence[0]
-            if sentence[0] == '(rdbg)'
-              cmd = @queue.pop
-              if cmd.is_a?(Proc)
-                cmd.call(result)
+        lines = []
+        begin
+          Timeout.timeout(10) do
+            while (line = read.gets)
+              lines.push(line)
+              debug_print line
+              case line.chomp
+              when '(rdbg)'
+                cmd = @queue.pop
+                if cmd.is_a?(Proc)
+                  cmd.call
+                  cmd = @queue.pop
+                end
+                write.puts(cmd)
+              when /INTERNAL_INFO:\s(.*)/
+                @internal_info = JSON.parse(Regexp.last_match(1))
+              when %r{Really quit\? \[Y/n\]}
                 cmd = @queue.pop
+                write.puts(cmd)
               end
-              write.puts(cmd)
-              quit = true if cmd == 'quit'
-            elsif sentence[0].include?('=>#0')
-              result = sentence[0]
             end
           end
-        end
-        read.expect(/.*\n/) do |sentence|
-          debug_print sentence[0]
-        end
-        read.expect(%r{(Really quit\? \[Y/n\])}) do |sentence|
-          debug_print sentence[0]
-          write.puts('y')
-        end
-        read.expect(/.*\n/) do |sentence|
-          debug_print sentence[0]
+        rescue Errno::EIO => e
+          if @queue.empty?
+            # result of `gets` return this exception in some platform
+            # https://github.com/ruby/ruby/blob/master/ext/pty/pty.c#L729-L736
+          else
+            p e
+            pp lines
+          end
+        rescue Timeout::Error => e
```

Comment:
On the timeout, the test should fail. Please try after merge it.

Reply:
Also, please try the timeout on your computer by adding `sleep 11` or something similar.